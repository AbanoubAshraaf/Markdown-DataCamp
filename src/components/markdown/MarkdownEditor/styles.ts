import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
    container: {
      flexGrow: 1,
      padding: 20,
      backgroundColor: '#fff',
    },
    textAreaPortrait: {
      height: 200,
      borderColor: '#ccc',
      borderWidth: 1,
      padding: 10,
      margin:10,
      borderRadius: 5,
      fontSize: 16,
    },
    textAreaLandscape: {
      flex: 1,
      borderColor: '#ccc',
      borderWidth: 1,
      padding: 10,
      margin:10,
      borderRadius: 5,
      fontSize: 16,
    },
  });

  