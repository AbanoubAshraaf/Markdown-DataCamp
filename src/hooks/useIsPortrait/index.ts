
import { useWindowDimensions } from 'react-native';
import { useState, useEffect } from 'react';

const useIsPortrait = () => {
  const { width, height } = useWindowDimensions();
  const [isPortrait, setIsPortrait] = useState(height > width);

  useEffect(() => {
    setIsPortrait(height > width);
  }, [width, height]);

  return isPortrait;
};

export default useIsPortrait;
