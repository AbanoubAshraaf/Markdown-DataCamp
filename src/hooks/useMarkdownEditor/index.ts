import { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Alert } from 'react-native';

const MARKDOWN_STORAGE_KEY = 'markdownText';

const useMarkdownEditor = () => {
  const [markdownText, setMarkdownText] = useState<string>('');
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    const loadMarkdown = async () => {
      try {
        const savedText = await AsyncStorage.getItem(MARKDOWN_STORAGE_KEY);
        if (savedText) {
          setMarkdownText(savedText);
        }
      } catch (error) {
        Alert.alert('Error', 'Failed to load saved markdown text.');
      } finally {
        setLoading(false);
      }
    };

    loadMarkdown();
  }, []);

  const saveMarkdown = async (text: string) => {
    try {
      await AsyncStorage.setItem(MARKDOWN_STORAGE_KEY, text);
    } catch (error) {
      Alert.alert('Error', 'Failed to save markdown text.');
    }
  };

  const handleReset = () => {
    setMarkdownText('');
  };

  return {
    markdownText,
    setMarkdownText,
    saveMarkdown,
    loading,
    handleReset,
  };
};

export default useMarkdownEditor;
