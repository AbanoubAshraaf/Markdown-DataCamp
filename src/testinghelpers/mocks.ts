export const mockedMarkDownText = 'markdown text'

export const mockedPortraitDimensions = {
    width: 400,
    height: 800,
    scale: 0,
    fontScale: 0,
};

export const mockedLandScapeDimensions = {
    width: 800,
    height: 400,
    scale: 0,
    fontScale: 0,
};