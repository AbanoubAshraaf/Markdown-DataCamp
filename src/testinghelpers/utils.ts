import { RenderResult } from "@testing-library/react-native";

interface Screen extends RenderResult {
    isDetached?: boolean;
}

export const createSnapShot = (screen: Screen) => {
    expect(screen).toMatchSnapshot();
  };