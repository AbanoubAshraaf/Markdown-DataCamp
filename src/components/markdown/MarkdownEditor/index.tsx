import React, { useCallback } from 'react';
import { ScrollView, View } from 'react-native';
import LoadingIndicator from 'src/components/basicComponent/LoadingIndicator';
import TextArea from 'src/components/basicComponent/TextArea';
import MarkdownViewer from 'src/components/markdown/MarkdownViewer';
import CustomButton from 'src/components/basicComponent/CustomButton';
import useMarkdownEditor from 'src/hooks/useMarkdownEditor';
import useIsPortrait from 'src/hooks/useIsPortrait';
import debounce from 'lodash.debounce';
import { styles } from './styles';

const MarkdownEditor: React.FC = () => {
  const { markdownText, setMarkdownText, saveMarkdown, loading, handleReset } = useMarkdownEditor();
  const isPortrait = useIsPortrait();

  const debouncedSave = debounce(saveMarkdown, 2000);

  const handleTextChange = useCallback((text: string) => {
    setMarkdownText(text);
    debouncedSave(text);
  }, []);

  if (loading) {
    return <LoadingIndicator />;
  }

  return (
    <ScrollView contentContainerStyle={styles.container}>
      <View style={{ flex: 1, flexDirection: isPortrait ? 'column' : 'row' }}>
        <TextArea
          value={markdownText}
          onChangeText={handleTextChange}
          placeholder="Enter your markdown here..."
          style={isPortrait ? styles.textAreaPortrait : styles.textAreaLandscape}
        />
        <MarkdownViewer markdownText={markdownText} />
      </View>
      <CustomButton title='Reset' onPress={handleReset} />
    </ScrollView>
  );
};



export default MarkdownEditor;
