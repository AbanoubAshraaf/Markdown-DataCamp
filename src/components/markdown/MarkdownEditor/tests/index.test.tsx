import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react-native';
import MarkdownEditor from '..';
import useMarkdownEditor from 'src/hooks/useMarkdownEditor';
import useIsPortrait from 'src/hooks/useIsPortrait';
import { mocked } from 'jest-mock';
import { mockedMarkDownText } from 'src/testinghelpers/mocks';
import { createSnapShot } from 'src/testinghelpers/utils';

jest.mock('src/hooks/useMarkdownEditor');
jest.mock('src/hooks/useIsPortrait');
jest.mock('@react-native-async-storage/async-storage', () => ({
    getItem: jest.fn(),
    setItem: jest.fn(),
}));

describe('MarkdownEditor', () => {
    afterEach(() => {
        jest.clearAllMocks();
        jest.clearAllTimers();
        jest.useRealTimers();
    });

    it('renders correctly, handles text change, and reset', async () => {
        const setMarkdownTextMock = jest.fn();
        const saveMarkdownMock = jest.fn();
        const handleResetMock = jest.fn();

        mocked(useMarkdownEditor).mockReturnValue({
            markdownText: '',
            setMarkdownText: setMarkdownTextMock,
            saveMarkdown: saveMarkdownMock,
            loading: false,
            handleReset: handleResetMock,
        });

        mocked(useIsPortrait).mockReturnValue(true);

        jest.useFakeTimers(); // Use fake timers

        const { getByPlaceholderText, getByText } = render(<MarkdownEditor />);

        const textArea = getByPlaceholderText('Enter your markdown here...');
        const resetButton = getByText('Reset');

        createSnapShot(screen);

        fireEvent.changeText(textArea, mockedMarkDownText);

        // Fast forward time for debounce
        jest.advanceTimersByTime(2000);
        expect(setMarkdownTextMock).toHaveBeenCalledWith(mockedMarkDownText);
        expect(saveMarkdownMock).toHaveBeenCalledWith(mockedMarkDownText);

        fireEvent.press(resetButton);
        expect(handleResetMock).toHaveBeenCalled();

        jest.useRealTimers(); // Switch back to real timers
    });

    it('shows loading indicator when loading is true', () => {
        mocked(useMarkdownEditor).mockReturnValue({
            markdownText: '',
            setMarkdownText: jest.fn(),
            saveMarkdown: jest.fn(),
            loading: true,
            handleReset: jest.fn(),
        });

        mocked(useIsPortrait).mockReturnValue(true);

        const { getByTestId } = render(<MarkdownEditor />);
        expect(getByTestId('loading-indicator')).toBeTruthy();

        // Snapshot the loading state
        createSnapShot(screen);
    });
});
