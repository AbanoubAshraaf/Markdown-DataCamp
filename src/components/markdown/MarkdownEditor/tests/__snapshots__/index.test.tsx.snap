// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownEditor renders correctly, handles text change, and reset 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "backgroundColor": "#fff",
      "flexGrow": 1,
      "padding": 20,
    }
  }
>
  <View>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <TextInput
        multiline={true}
        onChangeText={[Function]}
        placeholder="Enter your markdown here..."
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "fontSize": 16,
            "height": 200,
            "margin": 10,
            "padding": 10,
          }
        }
        value=""
      />
      <RCTScrollView
        style={
          {
            "backgroundColor": "#f0f0f0",
            "borderRadius": 5,
            "flex": 1,
            "margin": 10,
            "padding": 10,
          }
        }
      >
        <View>
          <View
            style={{}}
          />
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        {
          "marginBottom": 20,
          "marginTop": 10,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Reset
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`MarkdownEditor shows loading indicator when loading is true 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    size="large"
    testID="loading-indicator"
  />
</View>
`;
